### PR TITLE
feat: create WebSocket Reverse Proxy

### PR DESCRIPTION
**Pull request by fortuna**
This is just a utility to help develop the client code for Shadowsocks over Websockets.

### CODE REVIEW

**Comment by fortuna** [comment]
Any thoughts?

**Comment by fortuna** [comment]
Ping

File: x/examples/ws2endpoint/main.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// Copyright 2023 Jigsaw Operations LLC
```
Comment (jyyi1) [inline x/examples/ws2endpoint/main.go:1]:
```suggestion
// Copyright 2024 Jigsaw Operations LLC
```

File: x/examples/ws2endpoint/main.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/x/config"
+	"golang.org/x/net/websocket"
+)
+
+func main() {
+	addrFlag := flag.String("localAddr", "localhost:8080", "Local proxy address")
+	transportFlag := flag.String("transport", "", "Transport config")
+	endpointFlag := flag.String("endpoint", "", "Address of the target endpoint")
+	pathPrefix := flag.String("path", "/", "Path where to run the Websocket forwarder.")
```
Comment (jyyi1) [inline x/examples/ws2endpoint/main.go:37]: *nit* Remove the dot to align with other flag description, or you may append the dot to other flags too.

```suggestion
	pathPrefix := flag.String("path", "/", "Path where to run the Websocket forwarder")
```

File: x/examples/ws2endpoint/main.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		log.Fatal("Must specify flag -endpoint")
+	}
+	endpoint := transport.StreamDialerEndpoint{Dialer: dialer, Address: *endpointFlag}
+
+	listener, err := net.Listen("tcp", *addrFlag)
+	if err != nil {
+		log.Fatalf("Could not listen on address %v: %v", *addrFlag, err)
+	}
+	defer listener.Close()
+	log.Printf("Proxy listening on %v", listener.Addr().String())
```
Comment (jyyi1) [inline x/examples/ws2endpoint/main.go:54]: IIRC, `log.Printf` won't append a new line at the end.

```suggestion
	log.Printf("Proxy listening on %v\n", listener.Addr().String())
```

File: x/examples/ws2endpoint/main.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log.Printf("Got request: %v", r)
```
Comment (jyyi1) [inline x/examples/ws2endpoint/main.go:57]:
```suggestion
		log.Printf("Got request: %v\n", r)
```

File: x/examples/ws2endpoint/main.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		handler := func(wsConn *websocket.Conn) {
+			targetConn, err := endpoint.ConnectStream(r.Context())
+			if err != nil {
+				log.Printf("Failed to upgrade: %v", err)
```
Comment (jyyi1) [inline x/examples/ws2endpoint/main.go:61]:
```suggestion
				log.Printf("Failed to upgrade: %v\n", err)
```

File: x/examples/ws2endpoint/main.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		if err := server.Serve(listener); err != nil && err != http.ErrServerClosed {
+			log.Fatalf("Error running web server: %v", err)
+		}
+	}()
+
+	// Wait for interrupt signal to stop the proxy.
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt)
+	<-sig
+	log.Print("Shutting down")
```
Comment (jyyi1) [inline x/examples/ws2endpoint/main.go:85]:
```suggestion
	log.Println("Shutting down")
```

File: x/examples/ws2endpoint/README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  "city": "New York City",
+  "region": "New York",
+  "country": "US",
+  ...
+  "timezone": "America/New_York",
+  "readme": "https://ipinfo.io/missingauth"
+}
+```
+
+You can expose your WebSockets on Cloudflare with [clourdflared](https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/do-more-with-tunnels/trycloudflare/). For example:
```
Comment (jyyi1) [inline x/examples/ws2endpoint/README.md:45]: Will this WebSocket endpoint be publicly accessible? If so, we might need to consider applying some security best practices, for example, maximum size, timeout, etc.

File: x/examples/ws2endpoint/main.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	defer listener.Close()
+	log.Printf("Proxy listening on %v", listener.Addr().String())
+
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log.Printf("Got request: %v", r)
+		handler := func(wsConn *websocket.Conn) {
+			targetConn, err := endpoint.ConnectStream(r.Context())
+			if err != nil {
+				log.Printf("Failed to upgrade: %v", err)
+				w.WriteHeader(http.StatusBadGateway)
```
Comment (jyyi1) [inline x/examples/ws2endpoint/main.go:62]: Shall we `return` after writing the error response?

File: x/examples/ws2endpoint/main.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+			}
+			defer targetConn.Close()
+			go func() {
+				io.Copy(targetConn, wsConn)
+				targetConn.CloseWrite()
+			}()
+			io.Copy(wsConn, targetConn)
+			wsConn.Close()
+		}
+		websocket.Server{Handler: handler}.ServeHTTP(w, r)
```
Comment (jyyi1) [inline x/examples/ws2endpoint/main.go:72]: Will the `handler` be terminated when we shutdown the outer HTTP server (i.e., the server that owns `w` and `r`)?

**Review by jyyi1 — APPROVED** [review]
Thanks! The logic LGTM, with some questions.

File: x/examples/ws2endpoint/README.md (hunk 7 above)
Comment (fortuna) [reply x/examples/ws2endpoint/README.md:45]: This is just an example. I added a note about authentication below.

File: x/examples/ws2endpoint/main.go (hunk 8 above)
Comment (fortuna) [reply x/examples/ws2endpoint/main.go:62]: Yes, good catch.

File: x/examples/ws2endpoint/main.go (hunk 9 above)
Comment (fortuna) [reply x/examples/ws2endpoint/main.go:73]: It's just a function. There's nothing to shut down.

**Review by jyyi1 — APPROVED** [review]

**Pull request merged**